app/rtc_compare.py

```diff
@@ -0,0 +1,386 @@
+    if is_attr:
+        # str_key is for attribute
+        path_attr, key_attr = str_key.split('\n')
+        val_1 = hdf5_obj_1[path_attr].attrs[key_attr]
+        val_2 = hdf5_obj_2[path_attr].attrs[key_attr]
+        # Force the datatype of val_1 and _2 to make use of numpy features
+        if not isinstance(val_1,np.ndarray):
+            val_1 = np.array(val_1)
+
+        if not isinstance(val_2,np.ndarray):
+            val_2 = np.array(val_2)
+
+    else:
+        # str_key is for dataset
+        val_1 = np.array(hdf5_obj_1[str_key])
+        val_2 = np.array(hdf5_obj_2[str_key])
+
+    shape_val_1 = val_1.shape
+    shape_val_2 = val_2.shape
+
+    if shape_val_1 != shape_val_2:
+        # Dataset or attribute shape does not match
+        print(f'    - Data shapes do not match. {shape_val_1} vs. {shape_val_2}')
+        return False
+
+    if len(shape_val_1)==0:
+        # Scalar value
+        print(f'    - 1st value: {val_1}')
+        print(f'    - 2nd value: {val_2}')
+
+        if issubclass(val_1.dtype.type, np.number) and\
+        issubclass(val_2.dtype.type, np.number):
+
+            # numerical array
+            return_val = np.array_equal(val_1, val_2, equal_nan=True)
+            if not return_val:
+                print('    - numerical scalar. Failed to pass np.array_equal()')
+            return return_val
+
+        # Not a numerical array
+        return_val = np.array_equal(val_1, val_2)
+        if not return_val:
+            print('    - non-numerical scalar. Failed to pass np.array_equal()')
+        return return_val
+
+    if len(shape_val_1)==1:
+        # 1d vector
+
+        # Dereference if val_1 and val_2 have HDF5 objstc reference.
+        # Convert the 1d numpy array into list to differentiate the comparison process
+        if 'shape' in dir(val_1[0]):
+            if isinstance(val_1[0], np.void) or\
+            ((len(val_1[0].shape) == 1) and (isinstance(val_1[0][0], h5py.h5r.Reference))):
+                # Example:
+                # attribute `REFERENCE_LIST` in
+                # /science/CSAR/RTC/grids/frequencyA/xCoordinates'
+                # attribute `DIMENSION_LIST` in
+                # /science/CSAR/RTC/grids/frequencyA/VH
+                list_val_1 = list(itertools.chain.from_iterable(val_1))
+                val_1_new = [None] * len(list_val_1)
+                for i_val, element_1 in enumerate(list_val_1):
+                    if isinstance(element_1, h5py.h5r.Reference):
+                        print('    - Object reference found.')
+                        val_1_new[i_val] = np.str_(hdf5_obj_1[element_1].name)
+                    else:
+                        val_1_new[i_val] = element_1
+                val_1 = np.array(val_1_new)
+
+        # Repeat the same process for `val_2`
+        if 'shape' in dir(val_2[0]):
+            if isinstance(val_2[0], np.void) or\
+            ((len(val_2[0].shape) == 1) and (isinstance(val_2[0][0], h5py.h5r.Reference))):
```

Comment:
Sorry, @seongsujeong . I still don't understand why you are checking `h5py.h5r.Reference` here an not at the beginning. If you did that at the beginning, you could check right away if the both references were pointing to the same HDF5 dataset. Also, you could check earlier if both data type match and that should simplify a lot your code.

Reply:
That is because of the structure of the variables with the object reference. I've found that they are `np.ndrray` of `np.ndarray`. For simple 1d arrays, `len(val_1[0].shape)` will give error because `val_1[0]` is scalar in that case.
The information below is how the attribute `DIMENSION_LIST` in `/science/CSAR/RTC/grids/frequencyA/VH` looks like.


```
>>> val_1
array([array([<HDF5 object reference>], dtype=object),
       array([<HDF5 object reference>], dtype=object)], dtype=object)
>>>
>>> val_1[0]
array([<HDF5 object reference>], dtype=object)
```

That means we cannot put `'shape' in dir(val_1[0])` and `(len(val_1[0].shape) == 1` in the same `if` statement. We have to check if the instance has `shape` attribute before accessing it.

Reply:
Uhm. Can't we check and convert each of these object references to strings, at the beginning of the function?

Reply:
> Uhm. Can't we check and convert each of these object references to strings, at the beginning of the function?

Sorry I've misunderstood your comment. I've moved the conversion routine right after loading the data from the HDF5 object. See my comment [here](https://github.com/opera-adt/RTC/pull/9/files#r984187729)